cmd/docker/docker.go

```diff
@@ -187,17 +191,58 @@ func setValidateArgs(dockerCli command.Cli, cmd *cobra.Command) {
 	})
 }
 
+func setupPluginSocket() (*net.UnixListener, error) {
+	uuid := uuid.Generate().String()
+	pluginCloseSocketAddr := net.UnixAddr{Name: "@docker_cli_" + uuid, Net: "unix"}
+	return net.ListenUnix("unix", &pluginCloseSocketAddr)
+}
+
 func tryPluginRun(dockerCli command.Cli, cmd *cobra.Command, subcommand string, envs []string) error {
 	plugincmd, err := pluginmanager.PluginRunCommand(dockerCli, subcommand, cmd)
 	if err != nil {
 		return err
 	}
 	plugincmd.Env = append(envs, plugincmd.Env...)
 
-	go func() {
-		// override SIGTERM handler so we let the plugin shut down first
-		<-appcontext.Context().Done()
-	}()
+	listener, err := setupPluginSocket()
+	// if we somehow failed to set up the socket to alert the plugin
+	// when the CLI process wants to exit, fallback to old behaviour
+	if err == nil {
+		defer listener.Close()
+		plugincmd.Env = append(plugincmd.Env, plugin.CliPluginCloseSocketEnvKey+"="+listener.Addr().String())
```

Comment:
Probably should be a sprintf.

Reply:
Not sure we should use a sprintf for a basic concatenation; sprintf definitely comes with an overhead

Reply:
For something that happens once during the lifetime of a process, I think the readability win is big, but I concede that it is strictly less efficient.

Still, premature optimization is often the greatest enemy of maintainability, even if this example is trivial.

Reply:
Although I can't currently imagine a scenario where it would be different from the `Name` we set up, I feel in general (and for future-proofing's sake, too) it's safer to use `listener.Addr().String()` than to assume it will always be that in all platforms we build for.